File: hcs_toolbox_py/lpc2hpc.py
"""
Upgrade Linear Proportional Control (LPC) to Homogeneous Proportional Control (HPC).

Original MATLAB: lpc2hpc.m

The upgrade:
    Linear:  u = K * x
    HPC:     u = K0 * x + nx^(1+mu) * (K - K0) * d(-ln(nx)) * x

where nx is the canonical homogeneous norm of x.
"""

import numpy as np
from scipy.linalg import sqrtm
from .block_con import block_con


def lpc2hpc(A, B, K):
    """
    Upgrade a linear proportional controller u = K*x to a Homogeneous
    Proportional Controller (HPC).

    Parameters
    ----------
    A : (n, n) ndarray
        System matrix.
    B : (n, m) ndarray
        Control matrix.
    K : (m, n) ndarray
        Linear feedback gain matrix.  A + B*K must be Hurwitz.

    Returns
    -------
    K0 : (m, n) ndarray
        Feedback gain of the linear component of HPC.
    G0 : (n, n) ndarray
        Matrix defining the generator of dilation: Gd = I + mu*G0.
    P : (n, n) ndarray
        Positive definite matrix defining the canonical homogeneous norm
        induced by sqrt(x'*P*x).
    mu_min : float
        Minimal admissible homogeneity degree (mu_min <= mu).
    mu_max : float
        Maximal admissible homogeneity degree (mu <= mu_max).
    """
    tol = 1e-5
    n = A.shape[0]
    m = B.shape[1]

    # --- Controllability check ---
    U_ctrb = B.copy()
    tA = A.copy()
    k_ctrb = 0
    if np.linalg.matrix_rank(U_ctrb) == n:
        k_ctrb = 1
    i = 1
    while (k_ctrb == 0) and (i <= n - 1):
        i += 1
        U_ctrb = np.hstack([U_ctrb, tA @ B])
        if np.linalg.matrix_rank(U_ctrb) == n:
            k_ctrb = i
        tA = tA @ A

    if k_ctrb == 0:
        raise ValueError("lpc2hpc: The system is not controllable.")

    if np.linalg.det(U_ctrb @ U_ctrb.T) < tol:
        print("Warning: The system is weakly controllable. "
              "Parameters may be badly tuned due to computation issues.")

    # --- Stability margin check ---
    rho = -max(np.real(np.linalg.eigvals(A + B @ K))) * 0.001
    if rho < tol:
        raise ValueError("lpc2hpc: The linear control system does not have "
                         "sufficient stability margin. Upgrade impossible.")

    # --- Special case: B is square and invertible ---
    if np.linalg.matrix_rank(B) == n:
        K0 = -B.T @ np.linalg.inv(B @ B.T) @ A
        K_out = -B.T @ np.linalg.inv(B @ B.T)
        G0 = np.eye(n)
        P = np.eye(n)
        # mu_min, mu_max can be arbitrary for fully actuated systems
        return K0, G0, P, -1.0, 1.0

    # --- Block decomposition ---
    T, nt = block_con(A, B)
    if isinstance(T, int) and T == 0:
        raise ValueError("lpc2hpc: Block decomposition failed.")

    Anew = T @ A @ np.linalg.inv(T)
    k = len(nt)

    # Build block row start indices (0-based)
    n_ind = [0]
    s = 0
    for i in range(k - 1):
        s += nt[i]
        n_ind.append(s)

    Bnew = T @ B
    B0 = Bnew[n_ind[-1]:n, :m]
    K0 = -B0.T @ np.linalg.inv(B0 @ B0.T) @ Anew[n_ind[-1]:n, :n]
    K0 = K0 @ T

    # --- Compute G0 ---
    vG0 = []
    for i in range(k):
        vG0.extend([(k - 1 - i)] * nt[i])
    G0 = -np.linalg.inv(T) @ np.diag(vG0) @ T

    # --- Compute P from Lyapunov equation ---
    # (A+B*K)'*P + P*(A+B*K) = -2*eye(n)
    Acl = A + B @ K
    I_n = np.eye(n)

    # Solve via Kronecker product:
    # kron(I, Acl') + kron(Acl', I)  *  vec(P) = -2*vec(I)
    W0 = np.kron(I_n, Acl.T) + np.kron(Acl.T, I_n)
    zet0 = -2.0 * I_n.flatten('F')
    v_P = np.linalg.solve(W0, zet0)
    P = v_P.reshape((n, n), order='F')

    # --- Compute admissible mu range ---
    sqrtP = sqrtm(P)
    inv_sqrtP = np.linalg.inv(sqrtP)
    M_sym = sqrtP @ G0 @ inv_sqrtP + inv_sqrtP @ G0.T @ sqrtP
    lambdas = np.real(np.linalg.eigvals(M_sym))
    lambda_min = np.min(lambdas)
    lambda_max = np.max(lambdas)

    if lambda_max > tol:
        mu_min = max(-1.0, -1.0 / lambda_max + tol)
    else:
        mu_min = -1.0

    if lambda_min < -tol:
        mu_max = min(1.0 / k, -1.0 / lambda_min)
    else:
        mu_max = 1.0 / k

    return K0, G0, P, mu_min, mu_max

File: HCS_Toolbox_ver02/lpc2hpc.m
function [K0, G0, P, mu_min, mu_max]=lpc2hpc(A,B,K)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% The function [K0, Gd, P, mu_min, mu_max]=lpc2hpc(A,B,K) upgrades 
%% the linear proportional controller (LPC) 
%%
%%            u=K*x         (x -system state)
%%
%% to Homogeneous Proportional Controller (HPC):     
%%                                                                      
%%   u=K0*x+nx^(1+mu)*(K-K0)*d(-log(nx))*x     (nx - homogeneous norm of x)          
%%                                                                       
%%
%% The input parameters:                                             
%%      A  - system matrix (n x n)                                       
%%      B  - control matrix (n x m)       
%%      K  - gain matrix (m x n) of LPC s.t. A+B*K is Hurwitz  
%%
%% The output parameters:                                             
%%      K0 - feedback gain (m x n) of linear component of HPC                                          
%%      G0 - matrix (n x n) that  defines the generator Gd=eye(n)+mu*G0  
%%           of the dilation d(s)=expm(s*Gd)                   
%%      P  - positive definite matrix which defines the canonical        
%%           homogeneous norm nx induced by the norm sqrt(x'*P*x) 
%%  mu_min - minimal admissible degree of HPC  (mu_min<=mu)  
%%  mu_max - maximal admissible degree of HPC  (mu<=mu_max)
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
K0=0; G0=0; P=0; mu_min=0; mu_max=0;
tol=0.00001; 

% check dimensions
sA=size(A); sB=size(B); sK=size(K);

if sA(1)~=sA(2) disp('Error: the system matrix must be square'); return;
end;
if sB(1)~=sA(2) disp('Error:  dimensions of system  and control matrices must agree');return;
end;
if sK(1)~=sB(2) disp('Error:  dimensions of control and gain matrices must agree');return;
end;
n=sB(1); m=sB(2);

%controllability check
U=[B];tA=A;
k=0;
if rank(U)==n k=1;
end;i=1;
while (k==0) && (i<=n-1)
    i=i+1;
    U=[U tA*B];
    if rank(U)==n k=i; end;
    tA=A*tA;
end;
if k==0 disp('Error: The system is not controllable'); return;
end;
    
if det(U*U')<tol 
   disp('Warning: The system is weakly controllable. Parameters may be badly tuned due to computations issues.');
end;    

%% check is the closed-loop system is close to an unstable
rho=-max(real(eig(A+B*K)))*0.001;
if rho<tol disp('Error: The linear control system does not have a sufficient stability margin. The upgrade is impossible.'); return; end;

if rank(B)==n 
   K0=-B'*inv(B*B')*A;
   K=-B'*inv(B*B'); Gd=eye(n);
   P=eye(n);Gd=P;
   return;
end;   
    
%%%%%%%%%%%%%%%%%%%%%
% Design of control parameters K0, G0 using block decomposition
%%%%%%%%%%%%%%%%%%
[T nt]=block_con(A,B);
if T==0 K0=0; K=0;P=0; Gd=0; 
    disp('Error: A block decomposition is impossible.'); 
      return; end;
  
Anew=T*A*inv(T);
k=size(nt,1); 
%creation of an array of indecies
s=1;
n_ind=[s];
for i=1:k-1
s=s+nt(i);
n_ind=[n_ind; s];
end;

Bnew=T*B;
B0=Bnew(n_ind(k):n,1:m);
K0=-B0'*inv(B0*B0')*Anew(n_ind(k):n,1:n);
A0=Anew+Bnew*K0;
K0=K0*T;

vG0=[];
for i=1:k
 vG0=[vG0 (k-i)*ones(1,nt(i))];   
end;
G0=-inv(T)*diag(vG0)*T;

%% computation of P from Lyapunov equation (A+B*K)'*P+P(A+B*K)=-2*eye(n)
I_n=eye(n);
W0=[kron(I_n, (A+B*K)')+kron((A+B*K)',I_n)];
zet0=-[2*I_n(:)];
v_P=linsolve(W0,zet0);
P=reshape(v_P(1:n^2),n,n); 

%% calulations of admissible mu
lambda_min=min(real(eig(sqrtm(P)*G0*inv(sqrtm(P))+inv(sqrtm(P))*G0'*sqrtm(P))));
lambda_max=max(real(eig(sqrtm(P)*G0*inv(sqrtm(P))+inv(sqrtm(P))*G0'*sqrtm(P))));
if lambda_max>tol mu_min=max(-1,-1/lambda_max+tol);
else mu_min=-1;
end;
if lambda_min<-tol mu_max=min(1/k, -1/lambda_min);
else mu_max=1/k;
end;

